pkg/fuzzer/fuzzer.go

```diff
@@ -0,0 +1,298 @@
+// Copyright 2024 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package fuzzer
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/google/syzkaller/pkg/hash"
+	"github.com/google/syzkaller/pkg/ipc"
+	"github.com/google/syzkaller/pkg/rpctype"
+	"github.com/google/syzkaller/prog"
+)
+
+type Fuzzer struct {
+	Config    *Config
+	Corpus    *Corpus
+	NewInputs chan rpctype.Input
+
+	ctx    context.Context
+	mu     sync.Mutex
+	stats  map[string]uint64
+	rnd    *rand.Rand
+	target *prog.Target
+
+	ct      *prog.ChoiceTable
+	ctMu    sync.Mutex // TODO: use RWLock.
+	ctProgs int
+
+	nextExec *priorityQueue[*Request]
+	nextJob  *priorityQueue[job]
+}
+
+func NewFuzzer(ctx context.Context, cfg *Config, rnd *rand.Rand,
+	target *prog.Target) *Fuzzer {
+	f := &Fuzzer{
+		Config:    cfg,
+		Corpus:    newCorpus(),
+		NewInputs: make(chan rpctype.Input),
+
+		ctx:    ctx,
+		stats:  map[string]uint64{},
+		rnd:    rnd,
+		target: target,
+
+		nextExec: makePriorityQueue[*Request](),
+		nextJob:  makePriorityQueue[job](),
+	}
+	go func() {
+		<-ctx.Done()
+		close(f.NewInputs)
+	}()
+	return f
+}
+
+type Config struct {
+	Logf           func(level int, msg string, args ...interface{})
+	Coverage       bool
+	FaultInjection bool
+	Comparisons    bool
+	Collide        bool
+	EnabledCalls   map[*prog.Syscall]bool
+	NoMutateCalls  map[int]bool
+	LeakChecking   bool
+	FetchRawCover  bool
+	Candidates     chan Candidate
+}
+
+type Request struct {
+	Prog         *prog.Prog
+	NeedCover    bool
+	NeedRawCover bool
+	NeedSignal   bool
+	NeedHints    bool // TODO: can both it and something above be true?
+	// Fields that are only relevant within pkg/fuzzer.
+	flags      ProgTypes
+	stat       string
+	result     *Result
+	resultCond *sync.Cond
+	timeout    chan struct{}
+}
+
+func (req *Request) sent() {
+	// Let's catch lost requests.
+	time.AfterFunc(15*time.Minute, func() {
+		select {
+		case req.timeout <- struct{}{}:
+		default:
+		}
+	})
+}
+
+func (req *Request) wait() <-chan *Result {
+	ch := make(chan *Result)
+	go func() {
+		req.resultCond.L.Lock()
+		for req.result == nil {
+			req.resultCond.Wait()
+		}
+		req.resultCond.L.Unlock()
+		ch <- req.result
+	}()
+	return ch
+}
+
+type Result struct {
+	Info *ipc.ProgInfo
+	Stop bool
```

Comment:
Currently we set it only when Context is closed, but syz-fuzzer/syz-manager won't do it. So it's only for tests, right? But it adds a some amount of complexity and clutter to all jobs (they all need to check it and return on every exec).
Can we make all jobs run to completion in tests instead?

Reply:
It's also useful as one of the mechanisms to prevent repetitive crashes -- we'd abort all running jobs whose programs were on a crashed VM. So I think it's useful to keep the code aware of it from the beginning.